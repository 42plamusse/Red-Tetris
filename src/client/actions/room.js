import socket from "../services/socket-api";
import {
  LEAVE_ROOM,
  UPDATE_ROOM,
  UPDATE_ROOMS,
  JOIN_ROOM
} from "../../constants/constants";

export const updateRoom = room => ({
  type: UPDATE_ROOM,
  room
});

export const joinRoom = roomName => {
  return dispatch => {
    socket.emit(JOIN_ROOM, roomName, response => {});
  };
};

export const leaveRoom = () => {
  return dispatch => {
    socket.emit(LEAVE_ROOM);
    dispatch(updateRoom({}));
    window.location.hash = "";
  };
};

export const updateRooms = rooms => ({
  type: UPDATE_ROOMS,
  rooms
});
