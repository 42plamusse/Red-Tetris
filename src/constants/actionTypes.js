export const JOIN_ROOM = "JOIN_ROOM";
export const LEAVE_ROOM = "LEAVE_ROOM";
export const UPDATE_ROOM = "UPDATE_ROOM";
export const UPDATE_ROOMS = "UPDATE_ROOMS";

export const NEW_PLAYER = "NEW_PLAYER";
export const UPDATE_PLAYER = "UPDATE_PLAYER";
export const DISCONNECT_PLAYER = "DISCONNECT_PLAYER";

export const CONNECTION = "connection";
export const DISCONNECT = "disconnect";

export const UPDATE_SPECTRES = "UPDATE_SPECTRES";

export const KEY_PRESSED = "KEY_PRESSED";

export const START_GAME = "START_GAME";
