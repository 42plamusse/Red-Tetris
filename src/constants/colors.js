export const MALUS = "#3d3d3d";
