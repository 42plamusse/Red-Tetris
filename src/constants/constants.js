export const MALUS = "#3d3d3d";

export const JOIN_ROOM = "JOIN_ROOM";

export const UPDATE_ROOM = "UPDATE_ROOM";

export const LEAVE_ROOM = "LEAVE_ROOM";

export const NEW_PLAYER = "NEW_PLAYER";

export const DISCONNECT_PLAYER = "DISCONNECT_PLAYER";

export const UPDATE_PLAYER = "UPDATE_PLAYER";

export const UPDATE_PLAYERS_LIST = "UPDATE_PLAYERS_LIST";

export const HANDLE_HASH = "HANDLE_HASH";

export const UPDATE_ROOMS = "UPDATE_ROOMS";

export const DELETE_PLAYER = "DELETE_PLAYER";

export const SHOW_TOAST = "SHOW_TOAST";

export const CONNECTION = "connection";

export const UPDATE_SPECTRES = "UPDATE_SPECTRES";

export const ADD_CHAT_MESSAGE = "ADD_CHAT_MESSAGE";
export const NEW_CHAT_MESSAGE = "NEW_CHAT_MESSAGE";

export const DISCONNECT = "disconnect";

export const KEY_PRESSED = "KEY_PRESSED";
export const START_GAME = "START_GAME";

export const enumKeys = {
  ARROW_LEFT: "ArrowLeft",
  ARROW_RIGHT: "ArrowRight",
  ARROW_DOWN: "ArrowDown",
  ARROW_UP: "ArrowUp",
  SPACE: "Space",
  ENTER: "Enter"
};
export const ARROW_LEFT = "ArrowLeft";
export const ARROW_RIGHT = "ArrowRight";
export const ARROW_DOWN = "ArrowDown";
export const ARROW_UP = "ArrowUp";
export const SPACE = "Space";
export const ENTER = "Enter";

export const UPDATE_BOARD = "UPDATE_BOARD";
export const MAX_PLAYER = 5;

export const LOBBY_ROOM = "LOBBY_ROOM";

export const LOG_LINE =
  "\n_________________________________________________________________________________________________________\n\n";
