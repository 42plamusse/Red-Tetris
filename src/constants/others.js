export const MAX_PLAYER = 10;

export const LOBBY_ROOM = "LOBBY_ROOM";

export const LOG_LINE =
  "\n_________________________________________________________________________________________________________\n\n";
