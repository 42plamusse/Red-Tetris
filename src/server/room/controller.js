import Game from "../game/class";
import Room from "./class";
import { isAlphaNumeric } from "../../utils/utils";
import {
  LOBBY_ROOM,
  UPDATE_ROOM,
  UPDATE_ROOMS,
  UPDATE_PLAYERS
} from "../../constants/constants";
import { isTSExpressionWithTypeArguments } from "@babel/types";

export const joinRoom = (roomName, callback, socket, io) => {
  console.log("[CALL] joinRoom");
  if (!isAlphaNumeric(roomName) || roomName.length > 12) {
    callback({
      status: "error",
      message: "Room name must be 1 to 12 alphanumeric characters long"
    });
  } else {
    const player = Game.findPlayer(socket.id);
    if (player && !player.room) {
      let room = Game.findRoom(roomName);
      if (!room) {
        room = new Room(roomName, player.id);
        Game.addRoom(room);
      }
      if (room && !room.isFull()) {
        room.addPlayer(player);
        player.room = room;
        socket.leave(LOBBY_ROOM);
        socket.join(room.name);
        io.in(room.name).emit(UPDATE_ROOM, {
          room: room.toObject()
        });
        io.emit(UPDATE_ROOMS, { rooms: Game.createPublicRoomsArray() });
        callback({ status: "success" });
        // console.log(room);
      } else {
        callback({ status: "error", message: "Room is full" });
      }
      console.log("[UPDATED] after joinRoom", Game);
    }
    callback({ status: "error", message: "Socket problem" });
  }
};

export const leaveRoom = (socket, io) => {
  console.log("[CALL] leaveRoom on : ");
  const player = Game.findPlayer(socket.id);
  if (player && player.room) {
    const room = player.room;
    if (room.playersCount > 1) {
      room.removePlayer(player.id);
      if (player.id === room.hostId) {
        room.updateHostId();
      }
    } else {
      room.clean();
      Game.removeRoom(room.name);
    }
    player.clean();
    io.in(room.name).emit(UPDATE_PLAYERS, {
      players: room.createPublicPlayersArray()
    });
    socket.leave(room.name);
    socket.join(LOBBY_ROOM);
    io.in(room.name).emit(UPDATE_ROOM, {
      room: room.toObject()
    });
    io.emit(UPDATE_ROOMS, { rooms: Game.createPublicRoomsArray() });
    console.log("[UPDATED] after leaveRoom", Game);
  }
};

const updateClient = (room, io) => {
  room.players.forEach(player => {
    player.updateBoard();
  });
  io.in(room.name).emit(UPDATE_PLAYERS, {
    players: room.createPublicPlayersArray()
  });
};

const handleInterval = (room, io) => {
  room.players.forEach(player => {
    if (!player.piece.moveDown(player.heap)) {
      player.updateHeap();
    }
  });
  updateClient(room, io);
};

export const startGame = (socket, io) => {
  const room = Game.findPlayer(socket.id).room;
  if (room.isStarted) return;
  room.isStarted = true;
  room.players.forEach(player => player.newPiece());
  updateClient(room, io);
  room.interval = setInterval(() => handleInterval(room, io), 1000);
  console.log(room.interval);
};
