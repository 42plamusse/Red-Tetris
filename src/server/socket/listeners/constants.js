export const CREATE_ROOM = "CREATE_ROOM";

export const NEW_PLAYER = "NEW_PLAYER";

export const FETCH_ROOMS = "FETCH_ROOMS";

export const HASH_CHANGED = "HASH_CHANGED";

export const NEW_ROOM_LIST = "NEW_ROOM_LIST";
