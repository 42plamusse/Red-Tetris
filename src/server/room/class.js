import { MAX_PLAYER_BATTLEROYAL, SOLO } from "../../constants/constants";

class Room {
  constructor(name) {
    this._name = name;
    this._players = [];
    this._playersCount = 0;
    this._stillInGameCounter = 0;
    this._interval = null;
    this._isStarted = false;
    this._isGameOver = false;
    this._pieces = [];
    this._spectres = [];
    this._mode = SOLO;
  }

  get name() {
    return this._name;
  }
  get players() {
    return this._players;
  }
  get playersCount() {
    return this._playersCount;
  }
  get isStarted() {
    return this._isStarted;
  }
  set isStarted(bool) {
    this._isStarted = bool;
  }
  get interval() {
    return this._interval;
  }
  set interval(interval) {
    this._interval = interval;
  }
  get pieces() {
    return this._pieces;
  }
  set pieces(pieces) {
    this._pieces = pieces;
  }
  get spectres() {
    return this._spectres;
  }
  set spectres(spectres) {
    this._spectres = spectres;
  }
  get stillInGameCounter() {
    return this._stillInGameCounter;
  }
  set stillInGameCounter(stillInGameCounter) {
    this._stillInGameCounter = stillInGameCounter;
  }
  get isGameOver() {
    return this._isGameOver;
  }
  set isGameOver(isGameOver) {
    this._isGameOver = isGameOver;
  }
  get mode() {
    return this._mode;
  }
  set mode(mode) {
    this._mode = mode;
  }

  clean() {
    if (this._interval) clearInterval(this._interval);
    this._players.forEach(player => player.clean());
    this._pieces = [];
    this._isStarted = false;
  }

  addPlayer(player) {
    this._playersCount++;
    this._players.push(player);
  }

  updateHost() {
    this._players[0].isHost = true;
  }

  removePlayer(playerId) {
    this._playersCount--;
    this._players = this._players.filter(player => player.id !== playerId);
    this._spectres = this._spectres.filter(
      spectre => spectre.playerId !== playerId
    );
  }

  findPlayer(playerId) {
    return this._players.find(player => player.id === playerId);
  }

  isFull() {
    return this._playersCount >= MAX_PLAYER_BATTLEROYAL;
  }

  extendPiecesList() {
    for (let i = 0; i < 10; i++) {
      this._pieces.push(Math.floor(Math.random() * 7));
    }
  }

  newGame() {
    this.clean();
    this.extendPiecesList();
    this._isStarted = true;
    this._isGameOver = false;
    this._stillInGameCounter = this._players.length;
    this._players.forEach(player => player.newGame());
    this.initSpectres();
  }

  endGame() {
    if (this._interval) clearInterval(this._interval);
    this._isGameOver = true;
    this._isStarted = false;
  }

  fillHeap(heap) {
    const spectre = heap.map(row => [...row]);
    for (let i = 1; i < spectre.length; i++) {
      for (const j in spectre[i]) {
        if (spectre[i - 1][j]) spectre[i][j] = spectre[i - 1][j];
      }
    }
    return spectre;
  }

  initSpectres() {
    this._spectres = this._players.map(player => {
      return {
        playerId: player.id,
        playerName: player.name,
        board: player.heap.map(row => [...row])
      };
    });
  }

  updateSpectre(playerId, heap) {
    this._spectres = this._spectres.map(spectre => {
      if (spectre.playerId === playerId) {
        spectre.board = this.fillHeap(heap);
        return spectre;
      } else return spectre;
    });
  }

  toObject() {
    return {
      name: this._name,
      playersCount: this._playersCount,
      isStarted: this._isStarted,
      isGameOver: this._isGameOver
    };
  }
}

export default Room;
