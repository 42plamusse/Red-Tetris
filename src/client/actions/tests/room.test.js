import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import { leaveRoom, updateRoom, updateRooms } from "../room";
import { UPDATE_ROOM, UPDATE_ROOMS } from "../../../constants/actionTypes";
import socket from "../../services/socket-api";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

beforeEach(done => {
  // Setup
  if (socket.disconnected) {
    socket.connect();
  }
  socket.on("connect", () => {
    done();
  });
});

afterEach(done => {
  //   Cleanup;
  if (socket.connected) {
    socket.disconnect();
  }
  done();
});

describe("room actions", () => {
  test("leaveRoom clean room and hash", done => {
    const store = mockStore({
      room: {
        name: "Bob",
        hostId: "ID",
        playersCount: 2,
        isStarted: false
      }
    });
    window.location.hash = "#test[bob]";
    store.dispatch(leaveRoom());
    setTimeout(() => {
      store.getActions().forEach(action => expect(action).toMatchSnapshot());
      expect(window.location.hash).toBe("");
      done();
    }, 50);
  });
  test("updateRooms create the corresponding action", () => {
    const rooms = [
      { name: "Bob", playersCount: 3 },
      { name: "Jean", playersCount: 1 }
    ];
    const expectedAction = {
      type: UPDATE_ROOMS,
      rooms
    };
    expect(updateRooms(rooms)).toEqual(expectedAction);
  });
  test("updateRoom create the corresponding action", () => {
    const room = {
      name: "Bob",
      hostId: "ID",
      playersCount: 2,
      isStarted: false
    };
    const expectedAction = {
      type: UPDATE_ROOM,
      room
    };
    expect(updateRoom(room)).toEqual(expectedAction);
  });
});
