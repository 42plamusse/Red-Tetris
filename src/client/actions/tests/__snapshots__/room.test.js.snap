// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`room actions leaveRoom clean room and hash 1`] = `
Object {
  "room": Object {},
  "type": "UPDATE_ROOM",
}
`;
